### PR TITLE
Fix issue with member access on tables

### LINKED ISSUE #28371: Member access on tables panics when there is no entry with the specified key
**Description:**
$title.

The spec says
> if it is table, then if c does not contain a member with key k, the result is (); otherwise, the result is the member of c with key k

**Steps to reproduce:**
```ballerina
type Employee record {
    readonly string name;
    int id;
};

public function main() {
    table<Employee> key(name) t = table [
        {name: "Amy", id: 1234},
        {name: "John", id: 4567}
    ];
    Employee? e = t["Kate"];
}
```

```cmd
$ bal run xyz.bal 

Compiling source
        xyz.bal

Running executable

error: {ballerina/lang.table}KeyNotFound {"message":"cannot find key 'Kate'"}
        at xyz:main(xyz.bal:11)
```

**Affected Versions:**
slalpha1

### LINKED ISSUE #28474: Invalid static type for table member access
**Description:**
The spec says for member access, where the static type of the container table is `table<R> key<K>;`
> the static type of member-access-expr is R?.

**Steps to reproduce:**
The following should result in a compilation error.
```ballerina
type Employee record {
    readonly string name;
    int id;
};

public function main() {
    table<Employee> key(name) t = table [
        {name: "Amy", id: 1234},
        {name: "John", id: 4567}
    ];
    Employee e = t["Kate"]; // type of RHS should be `Employee?`
}
```

**Related Issues:**
https://github.com/ballerina-platform/ballerina-lang/issues/28371

**Affected Versions:**
slalpha1

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
Fix spec deviation in member access on tables where it panics when there is no entry with the specified key.
According to the spec,
`if it is table, then if c does not contain a member with key k, the result is (); otherwise, the result is the member of c with key k`

Fixes #28371
Fixes #28474

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/xml-type-table-key.bal [hunk 1]
```diff
@@ -83,14 +83,16 @@ function testMemberAccessWithInvalidXMLRecordKey() {
...
-    Customer customer = customerTable[xml `<id>245</id>`];
+    Customer? customer = customerTable[xml `<id>245</id>`];
+    assertEquality((), customer);
 }
 
 function testMemberAccessWithInvalidXMLMultiKey() {
     table<Customer> key(id, name) customerTable = table [{id: xml `<id>123</id>`, name: {fname: "Sanjiva", lname: "Weerawarana"}, address: "Sri Lanka" },
                                                    {id: xml `<id>234</id>`, name: {fname: "James" , lname: "Clark"}, address: "Thailand" }];
 
-    Customer customer = customerTable[xml `<id>245</id>`, {fname: "Sanjiva", lname: "Weerawarana"}];
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/table/xml-type-table-key.bal:93]: This should have (and should) result in a compilation error, right? The static type of the member access here should be `Customer?`. 

>  If T is a subtype of table<R>, then the static type of member-access-expr is R?.

I don't think we've fixed that yet. Shall we fix it and add negative tests?

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/table/xml-type-table-key.bal:93]: Created an issue for this too - https://github.com/ballerina-platform/ballerina-lang/issues/28474.

Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/table/xml-type-table-key.bal:93]: Fixed in dffc78b

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -6794,9 +6794,9 @@ private BType checkIndexAccessExpr(BLangIndexBasedAccess indexBasedAccessExpr) {
                     return symTable.semanticError;
                 }
             }
-
-            indexBasedAccessExpr.originalType = tableType.constraint;
-            actualType = tableType.constraint;
+            BType constraint = tableType.constraint;
+            indexBasedAccessExpr.originalType = constraint;
+            actualType = addNilForNillableAccessType(constraint);;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6799]: @lasinicl, as discussed let's fix `indexBasedAccessExpr.originalType` for nested access similar to https://github.com/ballerina-platform/ballerina-lang/pull/28322/files#diff-f4fbb6e8ffab357b74e188781ad3f6e3235f8e48b64fba8dfb6ad72bf9ccaf93R6696.

Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed in 1d1917e

Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed in 1d1917e

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -6800,9 +6800,10 @@ private BType checkIndexAccessExpr(BLangIndexBasedAccess indexBasedAccessExpr) {
                     return symTable.semanticError;
                 }
             }
-
-            indexBasedAccessExpr.originalType = tableType.constraint;
-            actualType = tableType.constraint;
+            BType constraint = tableType.constraint;
+            indexBasedAccessExpr.originalType = indexBasedAccessExpr.leafNode || !nillableExprType ? constraint :
+                    types.getTypeWithoutNil(constraint);;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6805]:
```suggestion
                    types.getTypeWithoutNil(constraint);
```

Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Shouldn't we first set `actualType = addNilForNillableAccessType(constraint);` and then use that in the checks here?

```suggestion
            indexBasedAccessExpr.originalType = indexBasedAccessExpr.leafNode || !nillableExprType ? actualType :
                    types.getTypeWithoutNil(actualType);
```

Please check with a sample where the intermediate access becomes nil.

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/typebynode/TypeByAccessExprTest.java [hunk 4]
```diff
@@ -91,9 +91,11 @@ public void visit(IndexedExpressionNode indexedExpressionNode) {
...
                     assertType(indexedExpressionNode, model, STRING);
                 } else {
-                    Optional<TypeSymbol> type = assertType(indexedExpressionNode, model, TYPE_REFERENCE);
-                    assertEquals(((TypeReferenceTypeSymbol) type.get()).name(), "Person");
-                    assertEquals(((TypeReferenceTypeSymbol) type.get()).typeDescriptor().typeKind(), RECORD);
+                    Optional<TypeSymbol> type = assertType(indexedExpressionNode, model, UNION);
+                    List<TypeSymbol> members = ((UnionTypeSymbol) type.get()).memberTypeDescriptors();
+                    assertEquals(members.get(0).typeKind(), TYPE_REFERENCE);
+                    assertEquals(((TypeReferenceTypeSymbol) members.get(0)).name(), "Person");
+                    assertEquals(((TypeReferenceTypeSymbol) members.get(0)).typeDescriptor().typeKind(), RECORD);
```
Comment (MaryamZi) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/typebynode/TypeByAccessExprTest.java:98]: Do we need to assert the nil member also here?

Looping in @pubudu91 too.

Comment (lasinicl) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/typebynode/TypeByAccessExprTest.java:98]: Added an assertion in 3b81168

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**